### PR TITLE
lint: replace flake8 with ruff

### PR DESCRIPTION
**Pull request by mattwang44**
## Types of changes

- [x] **Other (introduce ruff)**

## Description

Introduce ruff and remove flake8

### CODE REVIEW

**Review by josix — APPROVED** [review]

**Review by SivanYeh — APPROVED** [review]
LGTM! Will merge after #1172 merged.

**Comment by mattwang44** [comment]
rebased

File: .git-blame-ignore-revs
Comment (josix) [inline .git-blame-ignore-revs:1]: nit: add some comments here to illustrate the usage of this file
I just borrow the content from [`pydata/pydata-sphinx-theme`](https://github.com/pydata/pydata-sphinx-theme/pull/713/files)

```
# When making commits that are strictly formatting/style changes, add the
# commit hash here, so git blame can ignore the change.
#
# For more details, see:
# https://git-scm.com/docs/git-config#Documentation/git-config.txt-blameignoreRevsFile
#
```

Comment (mattwang44) [reply .git-blame-ignore-revs]: modified, thx

**Pull request merged**
